quinn/src/incoming_connection.rs

```diff
@@ -0,0 +1,163 @@
+use std::{
+    fmt,
+    future::{Future, IntoFuture},
+    net::{IpAddr, SocketAddr},
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use bytes::BytesMut;
+use proto::ConnectionError;
+use thiserror::Error;
+
+use crate::{
+    connection::{Connecting, Connection},
+    endpoint::EndpointRef,
+};
+
+/// An incoming connection for which the server has not yet begun its part of the handshake
+#[must_use = "futures/streams/sinks do nothing unless you `.await` or poll them"]
+pub struct IncomingConnection(Option<State>);
+
+struct State {
+    inner: proto::IncomingConnection,
+    endpoint: EndpointRef,
+    response_buffer: BytesMut,
+}
+
+impl IncomingConnection {
+    pub(crate) fn new(
+        inner: proto::IncomingConnection,
+        endpoint: EndpointRef,
+        response_buffer: BytesMut,
+    ) -> Self {
+        Self(Some(State {
+            inner,
+            endpoint,
+            response_buffer,
+        }))
+    }
+
+    /// The local IP address which was used when the peer established
+    /// the connection
+    pub fn local_ip(&self) -> Option<IpAddr> {
+        self.0.as_ref().unwrap().inner.local_ip()
+    }
+
+    /// The peer's UDP address
+    pub fn remote_address(&self) -> SocketAddr {
+        self.0.as_ref().unwrap().inner.remote_address()
+    }
+
+    /// Whether the socket address that is initiating this connection has been validated
+    ///
+    /// This means that the sender of the initial packet has proved that they can receive traffic
+    /// sent to `self.remote_address()`.
+    pub fn remote_address_validated(&self) -> bool {
+        self.0.as_ref().unwrap().inner.remote_address_validated()
+    }
+
+    /// Attempt to accept this incoming connection (an error may still occur)
+    pub fn accept(mut self) -> Result<Connecting, ConnectionError> {
```

Comment:
Could we simplify the API by returning `Connecting` unconditionally, but constructing it such that it immediately yields a `ConnectionError` when awaited? That seems like it'd reduce the number of distinct error paths in downstream code, and reduce churn. For example, the change to the accept loop in quinn/examples/server.rs, which introduces a regression where accept errors aren't logged, would be avoided.

Reply:
On second thought, that's just `.await` again. No harm in being a little more fine-grained with the explicit API, I suppose!